### PR TITLE
Moving the consumer to typescript

### PR DESCRIPTION
**Pull request by iPromKnight**
Wip.. Her for observability right now.
Not much left to go.

### CODE REVIEW

File: src/node/consumer/src/lib/models/configuration/cache_config.ts
Comment (sleeyax) [inline src/node/consumer/src/lib/models/configuration/cache_config.ts:3]: Like my comments above, and from a functional programming POV, this doesn't *need* to be a class and could be a plain and simple JS object instead. The `MONGO_URI` getter could then be converted to a utility function which accepts the config as a parameter. I'll just leave this here as I'm not sure what direction is favored.

File: src/node/consumer/src/lib/services/extension_service.ts
Comment (sleeyax) [inline src/node/consumer/src/lib/services/extension_service.ts:69]: formatting

```suggestion

```

File: src/node/consumer/src/repository/models/content.ts
Comment (sleeyax) [inline src/node/consumer/src/repository/models/content.ts:9]: What's `declare` for?

File: src/node/consumer/tsconfig.json
Comment (sleeyax) [inline src/node/consumer/tsconfig.json:20]: tsx, which uses esbuild under the hood, doesn't support [emitDecoratorMetadata](https://www.npmjs.com/package/tsx#does-it-have-any-limitations). Do you think this could cause issues?

**Review by sleeyax — CHANGES_REQUESTED** [review]
Initial review, mostly questions and suggestions for improvement. Code style and personal preferences are to be discussed further with the team. 

A few more things I would like to mention:
- There's an inconsistency in file naming. Some files are named in  `camelCase`and others in `snake_case`. It doesn't really matter which one we pick but it should be [consistent](https://typescript-eslint.io/rules/naming-convention/). 
- I'm not particularly fond of the 'services' approach; it seems like a concept directly borrowed from C# and applied to TypeScript :sweat_smile:

**Comment by purple-emily** [comment]
Damn sleeyax! Nice

**Comment by iPromKnight** [comment]
> Initial review, mostly questions and suggestions for improvement. Code style and personal preferences are to be discussed further with the team. 
> 
> A few more things I would like to mention:
> - There's an inconsistency in file naming. Some files are named in  `camelCase`and others in `snake_case`. It doesn't really matter which one we pick but it should be [consistent](https://typescript-eslint.io/rules/naming-convention/). 
> - I'm not particularly fond of the 'services' approach; it seems like a concept directly borrowed from C# and applied to TypeScript :sweat_smile: 

Yeah totally - I'm going through renaming them as I go - though ti'd caught that one
I was thinking snake_case for file names, and pascalCase for imports?

File: src/node/consumer/src/jobs/processTorrents.ts
Comment (iPromKnight) [reply src/node/consumer/src/jobs/processTorrents.ts:20]: we will just make it optional
No need to assign it, as createdAt is automatically assigned by sequelize on writing. Its not largely important for the createdAt to be set at the start of the processing pipeline, more so when its written I think and is data complete

File: src/node/consumer/src/lib/helpers/boolean_helpers.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/helpers/boolean_helpers.ts:2]: Function it is then - anything that reduces code size reduces container size 😄 
Every little helps

File: src/node/consumer/src/lib/models/configuration/cache_config.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/models/configuration/cache_config.ts:3]: Good call
I'll convert to a constant exported object and move the get function to the cache service.
I think its important to at least keep these in their own model files, as the config file was getting pretty busy, and separately these are easier to maintain

File: src/node/consumer/src/repository/models/content.ts
Comment (iPromKnight) [reply src/node/consumer/src/repository/models/content.ts:9]: declare has to be using when using Sequelize in typescript, because otherwise the properties mask the properties sequelize generate

From here: https://sequelize.org/docs/v6/other-topics/typescript/

File: src/node/consumer/tsconfig.json
Comment (iPromKnight) [reply src/node/consumer/tsconfig.json:20]: Strange - it definitely generates it in the transpiled code

```js
// src/repository/models/torrent.ts
var Torrent = class extends import_sequelize_typescript3.Model {
  contents;
  files;
  subtitles;
};
__name(Torrent, "Torrent");
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.STRING(64), primaryKey: true })
], Torrent.prototype, "infoHash", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.STRING(32), allowNull: false })
], Torrent.prototype, "provider", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.STRING(512) })
], Torrent.prototype, "torrentId", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.STRING(512), allowNull: false })
], Torrent.prototype, "title", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.BIGINT })
], Torrent.prototype, "size", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.STRING(16), allowNull: false })
], Torrent.prototype, "type", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.DATE, allowNull: false })
], Torrent.prototype, "uploadDate", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.SMALLINT })
], Torrent.prototype, "seeders", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.STRING(8e3) })
], Torrent.prototype, "trackers", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.STRING(4096) })
], Torrent.prototype, "languages", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.STRING(16) })
], Torrent.prototype, "resolution", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.BOOLEAN, allowNull: false, defaultValue: false })
], Torrent.prototype, "reviewed", 2);
__decorateClass([
  (0, import_sequelize_typescript3.Column)({ type: import_sequelize_typescript3.DataType.BOOLEAN, allowNull: false, defaultValue: false })
], Torrent.prototype, "opened", 2);
__decorateClass([
  (0, import_sequelize_typescript3.HasMany)(() => Content, { foreignKey: "infoHash", constraints: false })
], Torrent.prototype, "contents", 2);
__decorateClass([
  (0, import_sequelize_typescript3.HasMany)(() => File, { foreignKey: "infoHash", constraints: false })
], Torrent.prototype, "files", 2);
Torrent = __decorateClass([
  (0, import_sequelize_typescript3.Table)({ modelName: "torrent", timestamps: true })
], Torrent);
```

File: src/node/consumer/src/jobs/processTorrents.ts
Comment (iPromKnight) [reply src/node/consumer/src/jobs/processTorrents.ts]: i'll defer to your suggestion - changed 😄

File: src/node/consumer/src/jobs/processTorrents.ts
Comment (iPromKnight) [reply src/node/consumer/src/jobs/processTorrents.ts]: added the rule

File: src/node/consumer/src/lib/enums/cache_types.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/enums/cache_types.ts]: Same for CS too -I've changed them

File: src/node/consumer/src/jobs/processTorrents.ts
Comment (sleeyax) [reply src/node/consumer/src/jobs/processTorrents.ts]: > we will just make it optional

Agreed, looking at it again this is the most sensible option.

File: src/node/consumer/tsconfig.json
Comment (sleeyax) [reply src/node/consumer/tsconfig.json:13]: Hmm maybe the docs where [they mention this caveat](https://www.npmjs.com/package/tsx#does-it-have-any-limitations) are outdated then, can't really explain it otherwise.

**Comment by sleeyax** [comment]
>I was thinking snake_case for file names, and pascalCase for imports?

Fine with me :)
 
> The services idea is kinda c# - i dunno what the naming convention would be in TS - provider? Or just torrent_entries etc Main concept is ensuring the single responsibility principle

Hmm I did a sloppy job at the end there explaining my concern. There's no issue with the services themselves, or how they are named, but just the way they are used. Exporting instances of classes isn't common in JS/TS. I suggest looking into introducing a dependency injection layer like [typedi](https://www.npmjs.com/package/typedi). I'll try to dedicate some time to this tomorrow.

**Comment by iPromKnight** [comment]
> > I was thinking snake_case for file names, and pascalCase for imports?
> 
> Fine with me :)
> 
> > The services idea is kinda c# - i dunno what the naming convention would be in TS - provider? Or just torrent_entries etc Main concept is ensuring the single responsibility principle
> 
> Hmm I did a sloppy job at the end there explaining my concern. There's no issue with the services themselves, or how they are named, but just the way they are used. Exporting instances of classes isn't common in JS/TS. I suggest looking into introducing a dependency injection layer like [typedi](https://www.npmjs.com/package/typedi). I'll try to dedicate some time to this tomorrow.

Absolutely agreed - felt meh when doing it
Looking at the repo for typedi, its not maintained anymore and everyone else has moved over to [inversify](https://www.npmjs.com/package/inversify)

**Comment by sleeyax** [comment]
Ehw why switch to yarn all of a sudden?

This PR is becoming much more than just a typescript rewrite, sneaking in other changes isn't nice.

**Comment by iPromKnight** [comment]
It's a small change. 

I can change it back. Parallel restores seem much faster with it is the only reason I switched.

Definitely not snuck in though, I made sure it was labelled in two prs lol

File: src/node/consumer/src/lib/helpers/boolean_helpers.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/helpers/boolean_helpers.ts]: expanded the boolean function to allow true/yes/1 and no/false/0

File: src/node/consumer/src/lib/services/metadata_service.ts
Comment (sleeyax) [inline src/node/consumer/src/lib/services/metadata_service.ts:28]: The default access modifier is `public` so you can omit that. And you don't need `=` to define a class method. It probably works because JS is weird but best stick to TS conventions.

```suggestion
    async getKitsuId(info: IMetaDataQuery): Promise<number | Error> => {
```

**Review by sleeyax — CHANGES_REQUESTED** [review]
Love the added unit tests, good work!

File: src/node/consumer/Dockerfile
Comment (iPromKnight) [reply src/node/consumer/Dockerfile:11]: You're right - i tested - thanks

File: src/node/consumer/esbuild.js
Comment (iPromKnight) [reply src/node/consumer/esbuild.js:52]: converted esbuild to ts, and ignored in eslint

File: .gitignore
Comment (iPromKnight) [reply .gitignore:407]: i've removed it - just to hide it from gitkraken betwene all my commits - its my local docker compose with all the ports open :P

File: src/node/consumer/src/lib/helpers/boolean_helpers.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/helpers/boolean_helpers.ts:2]: cheers

File: src/node/consumer/src/lib/interfaces/cache_service.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/interfaces/cache_service.ts:8]: Ant the type def for CacheMethod is any function because any function can be wrapped - wasn't sure how to represent

File: src/node/consumer/src/lib/interfaces/season_episode_map.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/interfaces/season_episode_map.ts:7]: Cheers - i'll make local the to file service file itself, non imported - itds the only place its used.

File: src/node/consumer/src/lib/models/composition_root.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/models/composition_root.ts:20]: With inversify we can go one step further and remove the constructors

```js
@injectable()
export class TorrentFileService implements ITorrentFileService {
    @inject(IocTypes.IMetadataService) metadataService: IMetadataService;
    @inject(IocTypes.ITorrentDownloadService) torrentDownloadService: ITorrentDownloadService;
    @inject(IocTypes.ILoggingService) logger: ILoggingService;
```

I've done this and updated all the tests to resolve objects from the ioc container instead of through the constructor

File: src/node/consumer/src/lib/repository/models/torrent.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/repository/models/torrent.ts:55]: these aren't actual part of the table

File: src/node/consumer/src/lib/services/metadata_service.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/services/metadata_service.ts:28]: I've done this for all public methods

All private methods i've had to leave as arrow functions, in order to avoid the context binding requirements we'd have to setup to call them with `this.`

Any of the injected services would be undefined without setting up bindings like this.someMethod.bind(this); for each method in the constructor

File: src/node/consumer/src/lib/repository/models/torrent.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/repository/models/torrent.ts:53]: Not required on relationships

File: src/node/consumer/src/lib/models/inversify_config.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/models/inversify_config.ts:1]: Yep - good call - i've moved to a setup folder outside of the lib folder, at the same level as main.ts

This means anything thats now in the lib folder, can technically be split to shared packages when we get to the other apps - addon, jacket addon etc

**Comment by iPromKnight** [comment]
> Love the added unit tests, good work!

Thanks 😄 
It definitely needed them

File: src/node/consumer/src/lib/models/composition_root.ts
Comment (sleeyax) [reply src/node/consumer/src/lib/models/composition_root.ts]: Oh that's way nicer! Thanks.

File: src/node/consumer/src/lib/services/metadata_service.ts
Comment (sleeyax) [reply src/node/consumer/src/lib/services/metadata_service.ts]: Hmm interesting. I'd like to explore this locally tomorrow, see if it can still be improved somehow. Not that it's bad per se, just looks a bit weird this way.

File: src/node/consumer/src/lib/helpers/boolean_helpers.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/helpers/boolean_helpers.ts]: Just fixed this in last commit, added tests for all helpers too, and expanded the config service tests to test for `yes/true/1/TRUE`  `no/false/0/FALSE` on one of the boolean values using the service

File: src/node/consumer/src/lib/services/metadata_service.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/services/metadata_service.ts]: came from this: https://github.com/Microsoft/TypeScript/wiki/'this'-in-TypeScript
When i had failing tests

Easiest way to test is the tracker service tests, as there is one public method and one private method, if you change the private method from an arrow function to a class method, the logger will be undefined unless you setup the binding context for it in the constructor

Comment (iPromKnight) [reply src/node/consumer/src/lib/services/metadata_service.ts]:
```js
import {ICacheService} from "@interfaces/cache_service";
import {ILoggingService} from "@interfaces/logging_service";
import {ITrackerService} from "@interfaces/tracker_service";
import {configurationService} from '@services/configuration_service';
import {IocTypes} from "@setup/ioc_types";
import axios, {AxiosResponse} from 'axios';
import {inject, injectable} from "inversify";

@injectable()
export class TrackerService implements ITrackerService {
    @inject(IocTypes.ICacheService) cacheService: ICacheService;
    @inject(IocTypes.ILoggingService) logger: ILoggingService;

    async getTrackers(): Promise<string[]> {
        return this.cacheService.cacheTrackers(this.downloadTrackers);
    }

    private async downloadTrackers(): Promise<string[]> {
        const response: AxiosResponse<string> = await axios.get(configurationService.trackerConfig.TRACKERS_URL);
        const trackersListText: string = response.data;
        // Trackers are separated by a newline character
        let urlTrackers = trackersListText.split("\n");
        // remove blank lines
        urlTrackers = urlTrackers.filter(line => line.trim() !== '');

        if (!configurationService.trackerConfig.UDP_ENABLED) {
            // remove any udp trackers
            urlTrackers = urlTrackers.filter(line => !line.startsWith('udp://'));

        }

        this.logger.info(`Trackers updated at ${Date.now()}: ${urlTrackers.length} trackers`);

        return urlTrackers;
    }
}
```

as opposed to

```js
import {ICacheService} from "@interfaces/cache_service";
import {ILoggingService} from "@interfaces/logging_service";
import {ITrackerService} from "@interfaces/tracker_service";
import {configurationService} from '@services/configuration_service';
import {IocTypes} from "@setup/ioc_types";
import axios, {AxiosResponse} from 'axios';
import {inject, injectable} from "inversify";

@injectable()
export class TrackerService implements ITrackerService {
    @inject(IocTypes.ICacheService) cacheService: ICacheService;
    @inject(IocTypes.ILoggingService) logger: ILoggingService;

    async getTrackers(): Promise<string[]> {
        return this.cacheService.cacheTrackers(this.downloadTrackers);
    }

    private downloadTrackers = async (): Promise<string[]> => {
        const response: AxiosResponse<string> = await axios.get(configurationService.trackerConfig.TRACKERS_URL);
        const trackersListText: string = response.data;
        // Trackers are separated by a newline character
        let urlTrackers = trackersListText.split("\n");
        // remove blank lines
        urlTrackers = urlTrackers.filter(line => line.trim() !== '');

        if (!configurationService.trackerConfig.UDP_ENABLED) {
            // remove any udp trackers
            urlTrackers = urlTrackers.filter(line => !line.startsWith('udp://'));

        }

        this.logger.info(`Trackers updated at ${Date.now()}: ${urlTrackers.length} trackers`);

        return urlTrackers;
    };
}
```

**Comment by purple-emily** [comment]
Wait it’s tagged as done!? Nice!! I’ll wait for @sleeyax to tell me what opinion I should have 😂

**Comment by iPromKnight** [comment]
Cheers. Aye one unresolved comment atm. Will see and then I'll rebase on main and resolve conflicts 😄

File: src/node/consumer/src/lib/services/metadata_service.ts
Comment (sleeyax) [reply src/node/consumer/src/lib/services/metadata_service.ts]: Binding looks cleaner imo, it's a native js feature whereas the other approach is kind of a hack to avoid using bind.

Comment (sleeyax) [reply src/node/consumer/src/lib/services/metadata_service.ts]: Not on pc rn so can't test but does `this.cacheService.cacheTrackers((...) => this.downloadTrackers(...)` work?

**Comment by sleeyax** [comment]
> Wait it’s tagged as done!? Nice!! I’ll wait for @sleeyax to tell me what opinion I should have 😂

It's marked as ready for review. I've reviewed 2x, maybe let us know your own opinion by reviewing too? I'm not gonna tell you what to think 😂

File: src/node/consumer/src/lib/services/metadata_service.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/services/metadata_service.ts]: It does but the problem with binding is there is no type safety inside the parameters passed to the function, not on the function itself so you can pass anything you want

An arrow isn't really a hack it's just a function with global context ^^ and is native too. 

I agree it would be nicer to use bind, but only if the caller also could adhere to type safety and not just once you hit the called method

Comment (iPromKnight) [reply src/node/consumer/src/lib/services/metadata_service.ts]: Using the arrow just preserves the lexical this of the owning class

Of we want type safety and want to bind then we have to bind the context in the method and imo that looks a bit messy

Like this:

'''
class Dog {
  public constructor (private name: string) { }
  private eat (this: Dog, food: string): void { // annotate this and food parameters
    console.log (`${this.name} is eating ${food}.`);
  }
  public run (): void {
    this.eat ('bone'); // OK
  }
}
'''

Comment (sleeyax) [reply src/node/consumer/src/lib/services/metadata_service.ts]: I was confused because the function looks like a class method but due to the assignment it's actually more like a property assignment, which is still valid JavaScript. If you separate this snippet and have a look at the transpiled code then the following typescript:

```ts
class Foo {
    private bar = (): number => {
        return 0;
    }

    private baz(): number {
        return 1;
    }
}
```

Becomes: 

```js
var Foo = /** @class */ (function () {
    function Foo() {
        this.bar = function () {
            return 0;
        };
    }
    Foo.prototype.baz = function () {
        return 1;
    };
    return Foo;
}());
```

Meh, it's probably fine to keep it the way you originally did then, only looks a bit weird in my book but ultimately it does what it's supposed to do.

File: src/node/consumer/esbuild.ts
Comment (sleeyax) [inline src/node/consumer/esbuild.ts:6]: Could be a record too.

```suggestion
type DevDependencies = Record<string, string>
```

File: src/node/consumer/esbuild.ts
Comment (sleeyax) [inline src/node/consumer/esbuild.ts:12]: Do we even need this anyways? I think we can get rid of the [external](https://esbuild.github.io/api/#external) exclusion list altogether because the dev dependencies aren't (and shouldn't be) imported in the code.

File: src/node/consumer/src/lib/services/metadata_service.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/services/metadata_service.ts]: Cool cheers :)

File: src/node/consumer/esbuild.ts
Comment (iPromKnight) [reply src/node/consumer/esbuild.ts:12]: I "think" its needed for build, as all the type definitions are used at build time i think - Might be wrong there though - I'll test when back on the computer tomorrow.

Comment (iPromKnight) [reply src/node/consumer/esbuild.ts]: Tested - you are right - we dont need this. removed

Comment (sleeyax) [reply src/node/consumer/esbuild.ts]: Nice, much better that way!

File: src/node/consumer/src/lib/helpers/boolean_helpers.ts
Comment (sleeyax) [inline src/node/consumer/src/lib/helpers/boolean_helpers.ts:8]: The if statement can be added to the switch like this:

```suggestion
        switch (value?.trim()?.toLowerCase()) {
            case undefined:
                return defaultValue;
            case 'true':
```

Looks a bit nicer imo.

**Review by sleeyax — APPROVED** [review]
A lot of points have been addressed and improved, thanks. I feel like it's getting ready to merge.

Could probably use some extra testing and - if possible - an additional review though, considering the size of this PR. It's easy to overlook something.

File: src/node/consumer/src/lib/helpers/boolean_helpers.ts
Comment (iPromKnight) [reply src/node/consumer/src/lib/helpers/boolean_helpers.ts]: Agreed - also rather than falling back to default value when something invalid is entered (in case of typo's in env vars), i've added an error message stating the boolean was rejected and outputting the possible values

**Comment by purple-emily** [comment]
@iPromKnight Anything else left to do? I'm not a competent JS/TS developer so I am not offering anything to this discussion :)

**Comment by iPromKnight** [comment]
> @iPromKnight Anything else left to do? I'm not a competent JS/TS developer so I am not offering anything to this discussion :)

I dont think we have anyone else to review it
I've looked through as much as I can to try to find anything missed, but its hard to review mistakes you've made yourself 😄 

Its ready to merge I think

**Comment by sleeyax** [comment]
Alright LGTM from my end too, we can still improve things in another PR later down the line :) To who goes the honors of hitting the merge button?

**Comment by iPromKnight** [comment]
In the words of SLJ

image

**Comment by sleeyax** [comment]
:clinking_glasses:

**Pull request merged**
